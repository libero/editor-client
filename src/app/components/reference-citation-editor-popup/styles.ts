import { makeStyles } from '@material-ui/core/styles';

export const useReferenceEditorStyles = makeStyles((theme) => ({
  refSelectionList: {
    width: 450,
    height: 270,
    overflow: 'auto'
  },
  refSelectionListItem: {
    padding: theme.spacing(1, 1.5),
    display: 'flex',
    lineHeight: '24px',
    margin: theme.spacing(0, 1.5),
    border: 'solid 1px rgba(0, 0, 0, 0.23)',
    borderBottom: 'none',
    '&:last-child': {
      borderBottom: 'solid 1px rgba(0, 0, 0, 0.23)',
      marginBottom: theme.spacing(1.5)
    }
  },
  filterField: {
    width: `calc(100% - ${theme.spacing(3)}px)`,
    margin: theme.spacing(1.5, 1.5, 1, 1.5)
  },
  hiddenIcon: {
    visibility: 'hidden'
  },
  refContent: {
    flexGrow: 1
  },
  refTick: {
    alignSelf: 'center',
    flexGrow: 0
  }
}));
