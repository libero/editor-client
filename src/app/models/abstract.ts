import { EditorState } from 'prosemirror-state';
import { DOMParser as ProseMirrorDOMParser } from 'prosemirror-model';
import { gapCursor } from 'prosemirror-gapcursor';

import { makeSchemaFromConfig } from 'app/models/utils';
import * as abstractConfig from 'app/models/config/abstract.config';
import { buildInputRules } from 'app/models/plugins/input-rules';
import { PlaceholderPlugin } from 'app/models/plugins/placeholder.plugin';
import { SelectionPlugin } from 'app/models/plugins/selection.plugins';

export function createAbstractState(content: Element): EditorState {
  const schema = makeSchemaFromConfig(abstractConfig.topNode, abstractConfig.nodes, abstractConfig.marks);
  const xmlContentDocument = document.implementation.createDocument('', '', null);

  if (content) {
    xmlContentDocument.appendChild(content);
  }

  return EditorState.create({
    doc: ProseMirrorDOMParser.fromSchema(schema).parse(xmlContentDocument),
    schema,
    plugins: [buildInputRules(), SelectionPlugin, gapCursor(), PlaceholderPlugin('Enter abstract')]
  });
}

export function createImpactStatementState(content: Element): EditorState {
  const schema = makeSchemaFromConfig(abstractConfig.topNode, abstractConfig.nodes, abstractConfig.marks);
  const xmlContentDocument = document.implementation.createDocument('', '', null);

  if (content) {
    xmlContentDocument.appendChild(content);
  }

  return EditorState.create({
    doc: ProseMirrorDOMParser.fromSchema(schema).parse(xmlContentDocument),
    schema,
    plugins: [buildInputRules(), SelectionPlugin, gapCursor(), PlaceholderPlugin('Enter impact statement')]
  });
}
