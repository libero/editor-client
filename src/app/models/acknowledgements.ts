import { DOMParser as ProseMirrorDOMParser } from 'prosemirror-model';
import { EditorState } from 'prosemirror-state';
import { Step } from 'prosemirror-transform';
import { gapCursor } from 'prosemirror-gapcursor';

import * as acknowledgementsConfig from './config/acknowledgements.config';
import { buildInputRules } from './plugins/input-rules';
import { makeSchemaFromConfig } from './utils';
import { PlaceholderPlugin } from './plugins/placeholder.plugin';
import { SelectionPlugin } from './plugins/selection.plugins';

export function createAcknowledgementsState(content?: Element, changeSteps?: [Step]): EditorState {
  if (content) {
    const ackTitle = content.querySelector('title');
    if (ackTitle) {
      content.removeChild(ackTitle);
    }
  }

  const schema = makeSchemaFromConfig(
    acknowledgementsConfig.topNode,
    acknowledgementsConfig.nodes,
    acknowledgementsConfig.marks
  );
  const xmlContentDocument = document.implementation.createDocument('', '', null);

  if (content) {
    xmlContentDocument.appendChild(content);
  }

  return EditorState.create({
    doc: ProseMirrorDOMParser.fromSchema(schema).parse(xmlContentDocument),
    schema,
    plugins: [buildInputRules(), SelectionPlugin, gapCursor(), PlaceholderPlugin('Enter acknowledgements')]
  });
}
