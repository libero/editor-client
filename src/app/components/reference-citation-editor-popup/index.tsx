import React, { useCallback, SyntheticEvent, useState } from 'react';
import ReactDOM from 'react-dom';
import { v4 as uuidv4 } from 'uuid';
import { Node as ProsemirrorNode } from 'prosemirror-model';
import { EditorView, NodeView } from 'prosemirror-view';
import { useSelector, Provider, useDispatch } from 'react-redux';
import { ThemeProvider } from '@material-ui/core/styles';
import { Popover, TextField, InputAdornment } from '@material-ui/core';
import { has, get } from 'lodash';
import CheckCircleIcon from '@material-ui/icons/CheckCircle';
import Interweave from 'interweave';
import AddIcon from '@material-ui/icons/Add';
import CancelIcon from '@material-ui/icons/Cancel';
import { NodeSelection, TextSelection } from 'prosemirror-state';

import { theme } from 'app/styles/theme';
import { Reference, ReferenceContributor } from 'app/models/reference';
import { getReferences } from 'app/selectors/manuscript.selectors';
import { store } from 'app/store';
import { stringifyEditorState } from 'app/utils/view.utils';
import { useReferenceEditorStyles } from 'app/components/reference-citation-editor-popup/styles';
import { ReactFCProps } from 'app/utils/types';
import { ModalContainer } from 'app/containers/modal-container';
import { ReferenceFormDialog } from 'app/containers/reference-form-dialog/reference-form-dialog';
import * as manuscriptActions from 'app/actions/manuscript.actions';

interface ReferenceCitationEditorPopupProps {
  editorView: EditorView | undefined;
  onClose(): void;
  onChange(ref: Reference): void;
  node?: ProsemirrorNode;
  y: number;
  x: number;
}

const getRefContributorName = (contributor: ReferenceContributor): string => {
  return get(contributor, 'groupName', get(contributor, 'lastName', ''));
};

const getRefListAuthorsNames = (ref: Reference) => {
  let authorNames = getRefContributorName(ref.authors[0]);
  if (ref.authors.length === 2) {
    authorNames += ` and ${getRefContributorName(ref.authors[1])}`;
  } else if (ref.authors.length > 2) {
    authorNames += ' et al.';
  }
  return authorNames;
};

const getRefListItemText = (ref: Reference) => {
  return [
    getRefListAuthorsNames(ref),
    get(ref.referenceInfo, 'year'),
    has(ref.referenceInfo, 'chapterTitle') ? stringifyEditorState(get(ref.referenceInfo, 'chapterTitle')) : undefined,
    has(ref.referenceInfo, 'articleTitle') ? stringifyEditorState(get(ref.referenceInfo, 'articleTitle')) : undefined,
    has(ref.referenceInfo, 'dataTitle') ? stringifyEditorState(get(ref.referenceInfo, 'dataTitle')) : undefined,
    has(ref.referenceInfo, 'source') ? stringifyEditorState(get(ref.referenceInfo, 'source')) : undefined,
    has(ref.referenceInfo, 'conferenceName')
      ? stringifyEditorState(get(ref.referenceInfo, 'conferenceName'))
      : undefined
  ]
    .filter(Boolean)
    .join('. ');
};

const getRefNodeText = (ref: Reference) => {
  return [getRefListAuthorsNames(ref), get(ref.referenceInfo, 'year')].filter(Boolean).join(', ');
};

const renderReferenceModal = (props: ReactFCProps<typeof ReferenceFormDialog>) => {
  return <ModalContainer title={'Reference'} params={props} component={ReferenceFormDialog} />;
};

export const ReferenceCitationEditorPopup: React.FC<ReferenceCitationEditorPopupProps> = (props) => {
  const { editorView, x, y, node, onClose, onChange } = props;
  const refs = useSelector(getReferences);
  const [filteredRefs, setFilteredRefs] = useState<Reference[]>(refs);
  const [filterValue, setFilterValue] = useState<string>('');
  const [isReferenceDialogShown, setReferenceDialogShown] = useState<boolean>(false);
  const classes = useReferenceEditorStyles();
  const dispatch = useDispatch();

  const refId = node.attrs['refId'];
  const handleFilterChange = useCallback(
    (event) => {
      const filterValue = event.target['value'].toLowerCase();
      setFilterValue(event.target['value']);
      setFilteredRefs(
        refs.filter((ref) => {
          return getRefListItemText(ref).toLowerCase().indexOf(filterValue) >= 0;
        })
      );
    },
    [refs, setFilterValue]
  );
  const handleClick = useCallback(
    (event: SyntheticEvent) => {
      event.preventDefault();
      event.stopPropagation();
      const newRefId = (event.currentTarget as HTMLLIElement).dataset['refId'];
      if (newRefId !== refId) {
        onChange(refs.find(({ id }) => id === newRefId));
      }
    },
    [onChange, refId, refs]
  );

  const handleAddReference = useCallback(
    (newReference) => {
      dispatch(manuscriptActions.addReferenceAction(newReference));
      setReferenceDialogShown(false);
      onChange(newReference);
    },
    [dispatch, onChange, setReferenceDialogShown]
  );

  const clearFilterField = useCallback(() => {
    setFilterValue('');
    setFilteredRefs(refs);
  }, [setFilterValue, setFilteredRefs, refs]);

  const openReferenceFormDialog = useCallback(() => {
    setReferenceDialogShown(true);
  }, []);

  const closeReferenceFormDialog = useCallback(() => {
    setReferenceDialogShown(false);
  }, []);

  if (!editorView) {
    return <></>;
  }

  return (
    <Popover
      open={true}
      onClose={onClose}
      anchorReference="anchorPosition"
      anchorPosition={{ top: y, left: x }}
      anchorOrigin={{
        vertical: 'top',
        horizontal: 'left'
      }}
      transformOrigin={{
        vertical: 'top',
        horizontal: 'left'
      }}
    >
      <TextField
        classes={{ root: classes.filterField }}
        InputLabelProps={{ shrink: true }}
        label="Filter list"
        value={filterValue}
        variant="outlined"
        InputProps={{
          endAdornment: (
            <InputAdornment position="end">
              <CancelIcon onClick={clearFilterField} classes={{ root: classes.clearFilterIcon }} />
            </InputAdornment>
          )
        }}
        onChange={handleFilterChange}
      />
      <ul className={classes.refSelectionList}>
        <li className={classes.refSelectionListItem} onClick={openReferenceFormDialog}>
          <AddIcon fontSize="small" classes={{ root: classes.addReferenceIcon }} /> Reference
        </li>
        {filteredRefs.map((ref) => (
          <li className={classes.refSelectionListItem} key={ref.id} data-ref-id={ref.id} onClick={handleClick}>
            <div className={classes.refContent}>
              <Interweave content={getRefListItemText(ref)} />
            </div>
            <div className={classes.refTick}>
              <CheckCircleIcon color="primary" className={ref.id !== refId ? classes.hiddenIcon : ''} />
            </div>
          </li>
        ))}
      </ul>
      {isReferenceDialogShown
        ? renderReferenceModal({
            reference: undefined,
            onAccept: handleAddReference,
            onCancel: closeReferenceFormDialog,
            onDelete: undefined
          })
        : undefined}
    </Popover>
  );
};

export class ReferenceCitationNodeView implements NodeView {
  dom?: HTMLAnchorElement;
  refEditorContainer: HTMLDivElement;
  nodeSelection: NodeSelection;

  constructor(private node: ProsemirrorNode, private view: EditorView) {
    this.dom = document.createElement('a');
    this.dom.style.cursor = 'pointer';
    this.dom.textContent = this.node.attrs.refText || '???';

    this.handleChange = this.handleChange.bind(this);
    this.close = this.close.bind(this);
  }

  selectNode() {
    this.dom.classList.add('ProseMirror-selectednode');
    this.nodeSelection = this.view.state.selection as NodeSelection;
    this.open();
  }

  open() {
    const { $from } = this.view.state.selection;
    const start = this.view.coordsAtPos($from.pos);

    this.refEditorContainer = this.view.dom.parentNode.appendChild(document.createElement('div'));
    this.refEditorContainer.style.position = 'absolute';
    this.refEditorContainer.style.zIndex = '10';

    const x = start.left;
    const y = start.bottom;

    ReactDOM.render(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <ReferenceCitationEditorPopup
            editorView={this.view}
            node={this.node}
            onClose={this.close}
            onChange={this.handleChange}
            x={x}
            y={y}
          />
        </ThemeProvider>
      </Provider>,
      this.refEditorContainer
    );
  }

  close() {
    this.dom.classList.remove('ProseMirror-selectednode');
    ReactDOM.unmountComponentAtNode(this.refEditorContainer);
    this.refEditorContainer.parentNode.removeChild(this.refEditorContainer);
  }

  handleChange(ref: Reference) {
    const { from, to } = this.nodeSelection;
    const schema = this.view.state.schema;
    const change = this.view.state.tr.replaceWith(
      from,
      to,
      schema.nodes['refCitation'].create({ refId: ref.id || uuidv4(), refText: getRefNodeText(ref) })
    );
    // due browser managing cursor position on focus and blur the cursor is sometimes reset to 0
    // to rectify this behaviour we move cursor back to before the citation
    // change.setSelection(new TextSelection(change.doc.resolve(this.nodeSelection.$from.pos)));
    this.view.dispatch(change);
    this.close();
  }
}
