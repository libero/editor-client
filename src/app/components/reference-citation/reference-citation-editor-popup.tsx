import React, { useCallback, SyntheticEvent, useState } from 'react';
import { EditorView } from 'prosemirror-view';
import { useSelector, useDispatch } from 'react-redux';
import { Node as ProsemirrorNode } from 'prosemirror-model';
import { TextField, InputAdornment, Popper, Paper, ClickAwayListener } from '@material-ui/core';
import CheckCircleIcon from '@material-ui/icons/CheckCircle';
import Interweave from 'interweave';
import AddIcon from '@material-ui/icons/Add';
import ClearIcon from '@material-ui/icons/Clear';
import { has, get } from 'lodash';

import { getReferences } from 'app/selectors/manuscript.selectors';
import { getRefListAuthorsNames, Reference } from 'app/models/reference';
import { useReferenceEditorStyles } from 'app/components/reference-citation/styles';
import * as manuscriptActions from 'app/actions/manuscript.actions';
import { stringifyEditorState } from 'app/utils/view.utils';
import { ReactFCProps } from 'app/utils/types';
import { ReferenceFormDialog } from 'app/containers/reference-form-dialog/reference-form-dialog';
import { ModalContainer } from 'app/containers/modal-container';

interface ReferenceCitationEditorPopupProps {
  editorView: EditorView | undefined;
  onClose(): void;
  anchorEl: HTMLElement;
  onChange(ref: Reference): void;
  node?: ProsemirrorNode;
}

const renderReferenceModal = (props: ReactFCProps<typeof ReferenceFormDialog>) => {
  return <ModalContainer title={'Reference'} params={props} component={ReferenceFormDialog} />;
};

const getRefListItemText = (ref: Reference) => {
  return [
    `${getRefListAuthorsNames(ref)}, ${get(ref.referenceInfo, 'year')}`,
    has(ref.referenceInfo, 'chapterTitle') ? stringifyEditorState(get(ref.referenceInfo, 'chapterTitle')) : undefined,
    has(ref.referenceInfo, 'articleTitle') ? stringifyEditorState(get(ref.referenceInfo, 'articleTitle')) : undefined,
    has(ref.referenceInfo, 'dataTitle') ? stringifyEditorState(get(ref.referenceInfo, 'dataTitle')) : undefined,
    has(ref.referenceInfo, 'source') ? stringifyEditorState(get(ref.referenceInfo, 'source')) : undefined,
    has(ref.referenceInfo, 'conferenceName')
      ? stringifyEditorState(get(ref.referenceInfo, 'conferenceName'))
      : undefined
  ]
    .filter(Boolean)
    .join('. ');
};

export const ReferenceCitationEditorPopup: React.FC<ReferenceCitationEditorPopupProps> = (props) => {
  const { editorView, node, onClose, onChange, anchorEl } = props;
  const refs = useSelector(getReferences);
  const [filteredRefs, setFilteredRefs] = useState<Reference[]>(refs);
  const [filterValue, setFilterValue] = useState<string>('');
  const [isReferenceDialogShown, setReferenceDialogShown] = useState<boolean>(false);
  const classes = useReferenceEditorStyles();
  const dispatch = useDispatch();

  const refId = node.attrs['refId'];
  const handleFilterChange = useCallback(
    (event) => {
      const filterValue = event.target['value'].toLowerCase();
      setFilterValue(event.target['value']);
      setFilteredRefs(
        refs.filter((ref) => {
          return getRefListItemText(ref).toLowerCase().indexOf(filterValue) >= 0 || refId === ref.id;
        })
      );
    },
    [refs, setFilterValue, refId]
  );
  const handleClick = useCallback(
    (event: SyntheticEvent) => {
      event.preventDefault();
      event.stopPropagation();
      const newRefId = (event.currentTarget as HTMLLIElement).dataset['refId'];
      if (newRefId !== refId) {
        onChange(refs.find(({ id }) => id === newRefId));
      } else {
        onChange(undefined);
      }
    },
    [onChange, refId, refs]
  );

  const handleAddReference = useCallback(
    (newReference) => {
      dispatch(manuscriptActions.addReferenceAction(newReference));
      setReferenceDialogShown(false);
      onChange(newReference);
    },
    [dispatch, onChange, setReferenceDialogShown]
  );

  const clearFilterField = useCallback(() => {
    setFilterValue('');
    setFilteredRefs(refs);
  }, [setFilterValue, setFilteredRefs, refs]);

  const openReferenceFormDialog = useCallback(() => {
    setReferenceDialogShown(true);
  }, []);

  const closeReferenceFormDialog = useCallback(() => {
    setReferenceDialogShown(false);
  }, []);

  if (!editorView) {
    return <></>;
  }

  return (
    <Popper open={true} anchorEl={anchorEl} style={{ zIndex: 1 }}>
      <Paper>
        <ClickAwayListener onClickAway={onClose} mouseEvent="onMouseUp">
          <div>
            <TextField
              classes={{ root: classes.filterField }}
              InputLabelProps={{ shrink: true }}
              label="Filter list"
              value={filterValue}
              variant="outlined"
              InputProps={{
                endAdornment: (
                  <InputAdornment position="end">
                    <ClearIcon
                      onClick={clearFilterField}
                      color="disabled"
                      classes={{ root: classes.clearFilterIcon }}
                    />
                  </InputAdornment>
                )
              }}
              onChange={handleFilterChange}
            />
            <ul className={classes.refSelectionList}>
              <li className={classes.refSelectionListItem} onClick={openReferenceFormDialog}>
                <AddIcon fontSize="small" classes={{ root: classes.addReferenceIcon }} /> Add Reference
              </li>
              {filteredRefs.map((ref) => (
                <li className={classes.refSelectionListItem} key={ref.id} data-ref-id={ref.id} onClick={handleClick}>
                  <div className={classes.refContent}>
                    <Interweave content={getRefListItemText(ref)} />
                  </div>
                  <div className={classes.refTick}>
                    <CheckCircleIcon color="primary" className={ref.id !== refId ? classes.hiddenIcon : ''} />
                  </div>
                </li>
              ))}
            </ul>
            {isReferenceDialogShown
              ? renderReferenceModal({
                  reference: undefined,
                  onAccept: handleAddReference,
                  onCancel: closeReferenceFormDialog,
                  onDelete: undefined
                })
              : undefined}
          </div>
        </ClickAwayListener>
      </Paper>
    </Popper>
  );
};
