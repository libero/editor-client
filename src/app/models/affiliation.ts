import { getTextContentFromPath } from 'app/models/utils';
import { BackmatterEntity } from 'app/models/backmatter-entity';
import { JSONObject } from 'app/types/utility.types';

export class Affiliation extends BackmatterEntity {
  label: string;
  institution: { name: string };
  address: { city: string };
  country: string;

  constructor(data?: JSONObject | Element) {
    super();
    this.createEntity(data);
  }

  public getDisplayName(): string {
    return [this.institution.name, this.address.city, this.country].filter(Boolean).join(', ');
  }

  public clone(): Affiliation {
    const affitiation = new Affiliation();
    affitiation._id = this._id;
    affitiation.label = this.label;
    affitiation.country = this.country;
    affitiation.institution = { name: this.institution.name };
    affitiation.address = { city: this.address.city };
    return affitiation;
  }

  protected fromXML(xml: Element): void {
    this._id = xml.getAttribute('id');
    this.label = getTextContentFromPath(xml, 'label');
    this.institution = {
      name: [
        getTextContentFromPath(xml, 'institution[content-type="dept"]'),
        getTextContentFromPath(xml, 'institution:not([content-type])')
      ]
        .filter(Boolean)
        .join(', ')
    };

    this.address = {
      city: getTextContentFromPath(xml, 'addr-line named-content[content-type="city"]')
    };
    this.country = getTextContentFromPath(xml, 'country');
  }

  protected fromJSON(json: JSONObject): void {
    this._id = json.id as string;
    this.label = json.label as string;
    this.country = json.country as string;
    this.institution = json.institution as { name: string };
    this.address = json.address as { city: string };
  }

  protected createBlank(): void {
    this.label = '';
    this.country = '';
    this.institution = { name: '' };
    this.address = { city: '' };
  }
}

export function createAffiliationsState(affiliationsXml: Element[]): Affiliation[] {
  return affiliationsXml.map((xml) => new Affiliation(xml));
}
