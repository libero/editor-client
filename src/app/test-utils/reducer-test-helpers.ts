import { EditorState } from 'prosemirror-state';
import { Schema } from 'prosemirror-model';

import { Manuscript } from 'app/models/manuscript';
import { getInitialHistory, getLoadableStateSuccess } from 'app/utils/state.utils';
import { ManuscriptHistoryState } from 'app/store';

export function givenState(overrides: Partial<Manuscript>): ManuscriptHistoryState {
  return getLoadableStateSuccess(
    getInitialHistory({
      title: createDummyEditorState(),
      abstract: createDummyEditorState(),
      authors: [],
      affiliations: [],
      keywordGroups: {},
      articleInfo: {
        dtd: '',
        articleType: '',
        articleDOI: ''
      },
      references: [],
      ...overrides
    })
  );
}

export function createDummyEditorState(): EditorState {
  return EditorState.create({
    schema: new Schema({
      nodes: {
        doc: { content: 'inline*' },
        text: { inline: true, group: 'inline' }
      }
    }),
    plugins: []
  });
}
