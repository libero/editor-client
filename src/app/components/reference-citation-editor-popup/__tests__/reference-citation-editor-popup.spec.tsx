import React from 'react';
import { create } from 'react-test-renderer';
import { EditorView } from 'prosemirror-view';
import { mount } from 'enzyme';

import configureMockStore from 'redux-mock-store';
import { createBodyState } from 'app/models/manuscript-state.factory';
import { ReferenceCitationEditorPopup } from 'app/components/reference-citation-editor-popup/index';
import { Provider } from 'react-redux';
import { createReferenceAnnotatedValue, Reference } from 'app/models/reference';
import { givenState } from 'app/test-utils/reducer-test-helpers';
import { Popover } from '@material-ui/core';
import { EditorState } from 'prosemirror-state';

jest.mock('@material-ui/core', () => ({
  Popover: ({ children }) => <div data-cmp="Popover">{children}</div>,
  IconButton: ({ children }) => <div data-cmp="IconButton">{children}</div>,
  InputAdornment: ({ children }) => <div data-cmp="InputAdornment">{children}</div>,
  Button: ({ children }) => <div data-cmp="Button">{children}</div>,
  TextField: ({ value, onChange, children }) => (
    <input onChange={onChange} value={value} data-cmp="TextField">
      {children}
    </input>
  )
}));

jest.mock('@material-ui/core/styles', () => {
  return {
    ThemeProvider: ({ children }) => <>{children}</>,
    createMuiTheme: jest.fn(),
    makeStyles: jest.requireActual('@material-ui/core/styles').makeStyles
  };
});

const ReferenceData = [
  {
    id: 'bib1',
    authors: [
      {
        firstName: 'V',
        lastName: 'Berk'
      }
    ],
    type: 'journal',
    referenceInfo: {
      year: '2012',
      source: stringToEditorState('Science'),
      articleTitle: stringToEditorState(
        'Molecular architecture and assembly principles of <italic>Vibrio cholerae</italic> biofilms'
      ),
      doi: '',
      pmid: '',
      elocationId: '',
      firstPage: '236',
      lastPage: '239',
      inPress: false,
      volume: '337'
    }
  },
  {
    id: 'bib2',
    authors: [
      {
        firstName: 'PK',
        lastName: 'Feyerabend'
      }
    ],
    type: 'book',
    referenceInfo: {
      year: '2010',
      source: stringToEditorState('The skull and brain'),
      chapterTitle: stringToEditorState('The skull and brain'),
      publisherLocation: 'London',
      publisherName: 'Verso',
      edition: '4th Edition',
      editors: [],
      doi: '',
      pmid: '',
      elocationId: '',
      firstPage: '0',
      lastPage: '0',
      inPress: false,
      volume: '0'
    }
  }
] as Reference[];

describe('ReferenceCitationEditor', () => {
  const mockStore = configureMockStore([]);

  it('should render component', () => {
    const el = document.createElement('main-text');
    el.innerHTML = `<p><xref ref-type="bibr" rid="bib5">Harmon (2019)</xref></p>`;
    const editorState = createBodyState(el);

    const viewContainer = document.createElement('div');
    const editorView = new EditorView(viewContainer, {
      state: editorState,
      dispatchTransaction: jest.fn()
    });

    const store = mockStore({
      manuscript: givenState({ references: ReferenceData })
    });

    const node = editorView.state.doc.firstChild.content.firstChild;

    const wrapper = create(
      <Provider store={store}>
        <ReferenceCitationEditorPopup
          editorView={editorView}
          onChange={jest.fn()}
          node={node}
          onClose={jest.fn()}
          x={100}
          y={200}
        />
      </Provider>
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should trigger apply event', () => {
    const handleClose = jest.fn();
    const handleChange = jest.fn();
    const el = document.createElement('main-text');
    el.innerHTML = `<p><xref ref-type="bibr" rid="bib5">Harmon (2019)</xref></p>`;
    const editorState = createBodyState(el);

    const viewContainer = document.createElement('div');
    const editorView = new EditorView(viewContainer, {
      state: editorState,
      dispatchTransaction: jest.fn()
    });

    const store = mockStore({
      manuscript: givenState({ references: ReferenceData })
    });

    const node = editorView.state.doc.firstChild.content.firstChild;

    const wrapper = mount(
      <Provider store={store}>
        <ReferenceCitationEditorPopup
          editorView={editorView}
          onChange={handleChange}
          node={node}
          onClose={handleClose}
          x={100}
          y={200}
        />
      </Provider>
    );

    wrapper.find('li').at(0).simulate('click');

    expect(handleChange).toHaveBeenCalledWith(ReferenceData[0]);
  });

  it('should trigger close event', () => {
    const handleClose = jest.fn();
    const handleChange = jest.fn();
    const el = document.createElement('main-text');
    el.innerHTML = `<p><xref ref-type="bibr" rid="bib5">Harmon (2019)</xref></p>`;
    const editorState = createBodyState(el);

    const viewContainer = document.createElement('div');
    const editorView = new EditorView(viewContainer, {
      state: editorState,
      dispatchTransaction: jest.fn()
    });

    const store = mockStore({
      manuscript: givenState({ references: ReferenceData })
    });

    const node = editorView.state.doc.firstChild.content.firstChild;

    const wrapper = mount(
      <Provider store={store}>
        <ReferenceCitationEditorPopup
          editorView={editorView}
          onChange={handleChange}
          node={node}
          onClose={handleClose}
          x={100}
          y={200}
        />
      </Provider>
    );

    wrapper.find(Popover).prop('onClose')(new Event('input'), 'backdropClick');

    expect(handleClose).toHaveBeenCalled();
  });
});

function stringToEditorState(xmlContent: string): EditorState {
  const el = document.createElement('div');
  el.innerHTML = xmlContent;
  return createReferenceAnnotatedValue(el);
}
