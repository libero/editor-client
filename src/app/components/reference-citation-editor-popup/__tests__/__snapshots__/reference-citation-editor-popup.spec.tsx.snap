// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReferenceCitationEditor should render component 1`] = `
<div
  data-cmp="Popover"
>
  <ul
    className="makeStyles-refSelectionList-1"
  >
    <li
      className="makeStyles-refSelectionListItem-2"
      data-ref-id="bib1"
      onClick={[Function]}
    >
      <div
        className="makeStyles-refContent-4"
      >
        <span>
          Berk V et al.. 2012. Molecular architecture and assembly principles of 
          <em>
            Vibrio cholerae
          </em>
           biofilms. Science
        </span>
      </div>
      <div
        className="makeStyles-refTick-5"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root makeStyles-hiddenIcon-3 MuiSvgIcon-colorPrimary"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 15l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
          />
        </svg>
      </div>
    </li>
    <li
      className="makeStyles-refSelectionListItem-2"
      data-ref-id="bib2"
      onClick={[Function]}
    >
      <div
        className="makeStyles-refContent-4"
      >
        <span>
          Feyerabend PK et al.. 2010. The skull and brain. The skull and brain
        </span>
      </div>
      <div
        className="makeStyles-refTick-5"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root makeStyles-hiddenIcon-3 MuiSvgIcon-colorPrimary"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 15l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
          />
        </svg>
      </div>
    </li>
  </ul>
</div>
`;
